projects/output-and-error-handling/client/main.go

```diff
@@ -0,0 +1,182 @@
+
+	return retryDuration, nil
+}
+
+func handleTooManyRequestsResponse(response *http.Response) (time.Duration, error) {
+	// parse the Retry-After Header string
+	retryDuration, err := parseRetryAfterHeader(response.Header.Get("Retry-After"))
+
+	// if we cannot parse the Retry-After Header string then return an error
+	if err != nil {
+		return 0, fmt.Errorf("[2] failed to get a valid delay from retry-after header")
+	}
+
+	// if we reach here we have a retry duration (of type time.Duration)
+	switch {
+	case retryDuration > 5*time.Second:
+		return 0, fmt.Errorf("[2] retry-after is more than 5 seconds")
+	case retryDuration > 0:
+		return retryDuration, nil
+	default:
+		return 0, fmt.Errorf("[2] retry-after condition not found")
+	}
+}
+
+// this is returning a responseBody, a retry duration in seconds, and an error
+// this breaks single responsibility... need to rethink it... suggestions?
+func handleStatusCode(response *http.Response) (string, time.Duration, error) {
+	// handle the various response status codes (extensible for more status codes)
+	switch response.StatusCode {
+	// 200
+	case http.StatusOK:
+		// parse the response body
+		responseBody, err := parseResponseBody(response)
+
+		// if we cannot parse the response body then return an error
+		if err != nil {
+			return "", 0, fmt.Errorf("[1] parseResponseBody failed: %w", err)
+		}
+
+		// otherwise return the parsed response body
+		return responseBody, 0, nil
+
+	// 429
+	case http.StatusTooManyRequests:
+		// get the retry duration
+		retryDuration, err := handleTooManyRequestsResponse(response)
+
+		// if we cannot get a valid retry duration then return an error
+		if err != nil {
+			return "", 0, fmt.Errorf("[1] handleTooManyRequestsResponse failed: %w", err)
+		}
+
+		// return a retry duration
+		return "", retryDuration, nil
+
+	// Other Status Codes
+	default:
+		return "", 0, fmt.Errorf("[1] unhandled response status code: %d", response.StatusCode)
+	}
+}
+
+func makeGetRequest(url string) (string, time.Duration, error) {
+	// try to make the get request
+	response, err := http.Get(url)
+
+	// if the request fails then return an error
+	if err != nil {
+		return "", 0, fmt.Errorf("[0] request failed: %w", err)
+	}
+
+	// we must Close the body to signal to Go that we are done with the Response
+	// so it can release all the allocated resources
+	// use the defer keyword early to close the body regardless of the outcome
+	defer response.Body.Close()
+
+	// send the response to the status code handler
+	responseBody, retryDuration, err := handleStatusCode(response)
+
+	// if there is an error
+	if err != nil {
+		return "", 0, fmt.Errorf("[0] handleStatusCode failed: %w", err)
+	}
+
+	// if there is a retry duration
+	if retryDuration > 0 {
+		return "", retryDuration, nil
+	}
+
+	// return the parsed response body
+	return responseBody, 0, nil
+}
+
+func main() {
+	// loop until a successful response or an unrecoverable error
+	for {
```

Comment:
In the event of retries, this code currently handles each response independently. This is fine, but I'm curious what you think the behaviour should be if the server repeatedly responds saying "sleep for 4 seconds"? What if it does this 100 times in a row?

Reply:
I have added some rudimentary retry logic with a max of 3 (which seems to be an agreed standard)